### DISCUSSION TITLE
Why does errorMap not have precedence over initial message?

### DISCUSSION THREAD

**Discussion by janvorwerk** [colinhacks/zod#2997] [Q&A] [answered]
Hello,

First, let me take the opportunity to thank you all for having created & maintaining this lib. I am coming from another world and it's impressive what Typescipt allows... Zod being its paramount IMHO. 🚀🔥

---
## Here is my question.

Where I struggle the most, is regarding i18n and customization of error messages.

I am wondering why on this line of [parseUtil.ts](https://github.com/colinhacks/zod/blob/05d33f22b1001f29d8c3d2b1ded16a089a535072/src/helpers/parseUtil.ts#L31), the code reads:

```
    message: issueData.message || errorMessage,
```
rather than the opposite:
```
    message: errorMessage || issueData.message,
```

Indeed, the error map, is what the dev provides and it should have precedence, shouldn't it?

---
## Background: my use-case

I have a password validation schema:

* the password and its confirmation have to match
* the password needs to meet a complexity rule

```typescript
// Schema for the single password (used by tRPC)
export const changePasswordSchema = z.object({
  newPassword: z
    .string()
    .min(8)
    .superRefine((password: string | undefined, ctx: RefinementCtx) => {
      const isComplex = isComplexEnough(password);
      if (!isComplex) {
        ctx.addIssue({
          code: "custom",
          message: "passwordComplexity",    // <====== here this is a "sub-code"
        });
      }
    }),
});
// Adding the password confirmation (used by React Form Hook)
export const changePasswordForm = changePasswordSchema
  .partial()
  .extend({
    confirmPassword: z.string().optional(),
  })
  .superRefine((values, ctx) => {
    if (values.newPassword !== values.confirmPassword) {
      ctx.addIssue({
        code: "custom",
        path: ["confirmPassword"],
        message: "passwordsMismatch",    // <====== here this is a "sub-code"
      });
    }
  });
```

This would work perfectly, be it for the line mentionned above 😢

Anything I should be doing differently?

Thanks for reading me 🙏

**Comment by JacobWeisenburger** [comment]
Thanks for the thorough write up of your issue. Could you please provide some sample data that shows the problem?

**Reply by janvorwerk** [reply]
Hi @JacobWeisenburger, thanks for your quick reply!

What exactly do you mean with "sample data"? Do you actually mean an open-source test case? Of course I could put something together if needed, but if that's not what you need, I'd rather spare the time 😉

**Reply by JacobWeisenburger** [reply]
good question. I mean the data that would be parsed by your schemas.

Example:
```ts
const schema = z.object( { foo: z.string() } )
const validData = { foo: 'bar' } // this would be helpful, but not needed in this case
const invalidData = { foo: 42 } // this is what I'm looking for
```

**Reply by janvorwerk** [reply]
Understood. However, in order to run it on your side, you'd need the function `isComplexEnough()` which I omitted for brievety. I guess you can implement it as a dummy version, just checking if the password contains a `z`.

So that would be:

```typescript
function isComplexEnough(password: string) {
  return password.indexOf("z") >= 0;
}
const validData = { newPassword: "1234567z", confirmPassword: "1234567z" }; // identical with a 'z'
const failsComplexity = { newPassword: "12345678", confirmPassword: "12345678" }; // identical but no 'z'
const failsMismatch = { newPassword: "1234567z", confirmPassword: "0123456z" }; // different with a 'z'
const failsBoth = { newPassword: "12345678", confirmPassword: "01234567" }; // different without a 'z'
const failsMinLength =  { newPassword: "12345", confirmPassword: "12345" }; // identical but too short
```

**Reply by JacobWeisenburger** [reply]
perfect, I'll take a look

**Reply by janvorwerk** [reply]
If you want the actual check... here is what I have in my code (freely addapted from [this](https://forum.codewithmosh.com/t/password-complexity-for-zod/23622)):

```typescript
const isUpperCase = (ch: string) => /[A-Z]/.test(ch);
const isLowercase = (ch: string) => /[a-z]/.test(ch);
const isNumber = (ch: string) => /\d/.test(ch);
const isComplexEnough = (password: string | undefined) => {
  if (!password) return;
  let countOfUpperCase = 0,
    countOfLowerCase = 0,
    countOfNumbers = 0,
    countOfSpecialChar = 0;
  for (let i = 0; i < password.length; i++) {
    const ch = password.charAt(i);
    if (isNumber(ch)) countOfNumbers++;
    else if (isUpperCase(ch)) countOfUpperCase++;
    else if (isLowercase(ch)) countOfLowerCase++;
    else countOfSpecialChar++;
  }
  return countOfLowerCase > 0 && countOfUpperCase > 0 && countOfSpecialChar > 0 && countOfNumbers > 0;
};
```

And for what it's worth, if anybody finds this thread some time, I found an (ugly!) workaround by providing the translated string back into the i18n lib... which happens to return the key when it's not found.

**Comment by JacobWeisenburger** [comment]
> Indeed, the error map, is what the dev provides and it should have precedence, shouldn't it?

I'm confused, you talked about an "error map", but you didn't use an error map in your code. Could you explain what you mean by "error map" and how it relates to your schemas?

**Reply by janvorwerk** [reply]
Indeed.... Here is where I set the error map. Currently, I am sharing it with all my forms (but it's a starting project). Eventually I may have several flavors.

It's a `useZodForm()` React hook.


```typescript
import { zodResolver } from "@hookform/resolvers/zod";
import { useTranslations } from "next-intl";
import type { FieldValues, UseFormProps, UseFormReturn } from "react-hook-form";
import { useForm } from "react-hook-form";
import type {
  ErrorMapCtx,
  Schema,
  ZodCustomIssue,
  ZodErrorMap,
  ZodInvalidArgumentsIssue,
  ZodInvalidDateIssue,
  ZodInvalidEnumValueIssue,
  ZodInvalidReturnTypeIssue,
  ZodInvalidStringIssue,
  ZodInvalidTypeIssue,
  ZodInvalidUnionIssue,
  ZodIssueOptionalMessage,
  ZodNotMultipleOfIssue,
  ZodTooBigIssue,
  ZodTooSmallIssue,
  ZodUnrecognizedKeysIssue,
} from "zod";
import { ZodIssueCode } from "zod";

/**
 * A wrapper for the useForm, dedicated to zod validation with i18n.
 * By default, it will return the original zod error (untranslated, and not
 * necessarily what we want).
 *
 * @param schema The zod schema of the form
 * @param props the standard react-hook-form useForm() parameter (w/o the zod resolver)
 * @returns the augmented hook
 */
export function useZodForm<
  TFieldValues extends FieldValues = FieldValues,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  TContext = any,
  TTransformedValues extends FieldValues | undefined = undefined,
>(
  schema: Schema<TFieldValues>,
  props?: UseFormProps<TFieldValues, TContext>,
): UseFormReturn<TFieldValues, TContext, TTransformedValues> {
  // We use our translation hook first,
  // so that it can be used in the functions below
  const t = useTranslations("validationErrors");
  // const ct = useTranslations("validationErrors.custom");

  // ***************** all ZOD types are handled here *****************
  // see https://github.com/colinhacks/zod/blob/master/ERROR_HANDLING.md#zodissuecode
  //
  function invalidType(error: ZodInvalidTypeIssue, ctx: ErrorMapCtx): string {
    return ctx.defaultError;
  }
  function unrecognizedKeys(error: ZodUnrecognizedKeysIssue, ctx: ErrorMapCtx): string {
    return ctx.defaultError;
  }
  function invalidUnion(error: ZodInvalidUnionIssue, ctx: ErrorMapCtx): string {
    return ctx.defaultError;
  }
  function invalidEnumValue(error: ZodInvalidEnumValueIssue, ctx: ErrorMapCtx): string {
    return ctx.defaultError;
  }
  function invalidArguments(error: ZodInvalidArgumentsIssue, ctx: ErrorMapCtx): string {
    return ctx.defaultError;
  }
  function invalidReturnType(error: ZodInvalidReturnTypeIssue, ctx: ErrorMapCtx): string {
    return ctx.defaultError;
  }
  function invalidDate(error: ZodInvalidDateIssue, ctx: ErrorMapCtx): string {
    return ctx.defaultError;
  }
  function invalidString(error: ZodInvalidStringIssue, ctx: ErrorMapCtx): string {
    return ctx.defaultError;
  }
  function tooSmall(error: ZodTooSmallIssue, ctx: ErrorMapCtx): string {
    if (error.type === "string") {
      if (!ctx.data) {
        return t("valueRequired");
      }
      return t("tooShortMustBeLongerThan", {
        minimum: error.minimum as number,
      });
    }
    return ctx.defaultError;
  }
  function tooBig(error: ZodTooBigIssue, ctx: ErrorMapCtx): string {
    return ctx.defaultError;
  }
  function notMultipleOf(error: ZodNotMultipleOfIssue, ctx: ErrorMapCtx): string {
    return ctx.defaultError;
  }
  function custom(error: ZodCustomIssue, ctx: ErrorMapCtx): string {
    // Does not work. see https://github.com/colinhacks/zod/discussions/2997
    // Should be:
    // const key = error.message as keyof IntlMessages["validationErrors"]["custom"];
    // return ct(key);
    return ctx.defaultError;
  }

  // Triage the error codes here
  function buildI18nErrorMessage(error: ZodIssueOptionalMessage, ctx: ErrorMapCtx): string {
    switch (error.code) {
      case ZodIssueCode.invalid_type:
        return invalidType(error, ctx);
      case ZodIssueCode.unrecognized_keys:
        return unrecognizedKeys(error, ctx);
      case ZodIssueCode.invalid_union:
        return invalidUnion(error, ctx);
      case ZodIssueCode.invalid_enum_value:
        return invalidEnumValue(error, ctx);
      case ZodIssueCode.invalid_arguments:
        return invalidArguments(error, ctx);
      case ZodIssueCode.invalid_return_type:
        return invalidReturnType(error, ctx);
      case ZodIssueCode.invalid_date:
        return invalidDate(error, ctx);
      case ZodIssueCode.invalid_string:
        return invalidString(error, ctx);
      case ZodIssueCode.too_small:
        return tooSmall(error, ctx);
      case ZodIssueCode.too_big:
        return tooBig(error, ctx);
      case ZodIssueCode.not_multiple_of:
        return notMultipleOf(error, ctx);
      case ZodIssueCode.custom:
        return custom(error, ctx);
      default:
        return ctx.defaultError;
    }
  }

  // Our error map
  const i18nErrorMap: ZodErrorMap = (error: ZodIssueOptionalMessage, ctx: ErrorMapCtx) => ({
    message: buildI18nErrorMessage(error, ctx),
  });

  // Last but not least, return the standard form hook...
  // ... except that the id is added as well as the name as the return of register()
  // This makes it more compliant to the rule that labels 'for' should refer to an ID
  // and not to a name (Chrome is complaining about that otherwise)
  const { register: original, ...rest } = useForm<TFieldValues>({
    resolver: zodResolver(schema, { errorMap: i18nErrorMap }),
  });
  return { ...rest, register: (name, opts) => ({ id: name, ...original(name, opts) }) };
}
```



And here is how I use it with the above schema (and it's inferred type):
```typescript
  const form = useZodForm<ChangePasswordForm>(changePasswordForm);
```

**Comment by JacobWeisenburger** [comment]
> Indeed, the error map, is what the dev provides and it should have precedence, shouldn't it?

Why would you want the error map to take precedence over the custom errors you have in `superRefine`?

**Reply by janvorwerk** [reply]
Because of i18n. The hook with the error map has access to the i18n (which is a react hook as well) but the schema is shared in several places and does not have access to the i18n strings.
In my design the schema is a shared object (a bit as a DB model), and the error map is related to the business application.

**Reply by JacobWeisenburger** [reply]
this seems to be more of a problem with `react-hook-form` not having a good integration to override error messages.

**Reply by janvorwerk** [reply]
From your previous question, I start understanding that you envision the superRefine to be so specific that it should always win. 

Not sure what you mean with your last comment though 🤔

Because however `react-hook-form` (or anything else) would integrate with my schema, if the `superRefine()` always win, it means the schema has the last word, period. The form lib cannot do anything.

So,  to me the "problem" is the fact that my schema is a standalone definition with no dependency to the i18n stuff.

My rationale for doing so, is that these schemas are shared between several apps/packages (including a React Native app where the i18n API is different). To me, it also feels good that all my business constraints and data definitions can be used w/o grabbing any dependencies. I also use `jsonb` columns in a PostgreSQL database, and here again the field definitions are specified and validated using these zod schemas 🙏. 

Did I get something wrong?

So I guess reversing the order to become `message: errorMessage || issueData.message` would be a nasty breaking change for other users! I think that I have my answer wrt the rationale of your choice now.

However, to be able to support my use case of no-dependency-schemas + i18n, I suppose you could simply add a sub-code to the `ZodCustomIssue:`
```
export interface ZodCustomIssue extends ZodIssueBase {
  code: typeof ZodIssueCode.custom;
  subCode?: string;
  params?: { [k: string]: any };
}
``` 

As I write this, I realize that I could hijack the params to include my sub-code there... and remove it from the real params (when I have some) prior to sending to the i18n function... it's not very clean, but still better than my current workaround.

Any thoughts?

**Reply by JacobWeisenburger** [reply]
That was a lot, but I think you are on the right track now.

**Reply by janvorwerk** [reply]
Yes, I tend to write too much, sorry.

Would you consider adding the `subCode`?

Indeed there are many zod standard errors, and one single `custom`. Adding a sub-code would allow the users to extend the custom and have their codes as well. Like password-mismatch is different from password-complexity. Why would these be squeezed into a single `custom` category?

**Reply by JacobWeisenburger** [reply, marked as answer]
> Would you consider adding the `subCode`?

Unfortunately, I'm not on the team responsible for adding new features. But you are always welcome to make an issue for this or a PR.

> Why would these be squeezed into a single `custom` category?

I wasn't on the team when this was decided so I don't have an answer for that. I totally agree that `error.code` should be customizable.

**Reply by janvorwerk** [reply]
OK, thanks for your answers (and your patience reading me 🥹)!

**Reply by JacobWeisenburger** [reply]
lol, it's all good. If given the choice between too much and too little, I'll pick the former.
